drivers/sdhc/sdhc_cdns.c

```diff
@@ -0,0 +1,301 @@
+/*
+ * Copyright (C) 2023 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT cdns_sdhc
+
+#include <zephyr/drivers/sdhc.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/reset.h>
+
+#include "sdhc_cdns_ll.h"
+
+#define SDHC_CDNS_DESC_SIZE	(1<<20)
+#define COMBOPHY_ADDR_MASK	0x0000FFFFU
+
+#define DEV_CFG(_dev)	((const struct sdhc_cdns_config *)(_dev)->config)
+#define DEV_DATA(_dev)	((struct sdhc_cdns_data *const)(_dev)->data)
+
+LOG_MODULE_REGISTER(sdhc_cdns, CONFIG_SDHC_LOG_LEVEL);
+
+/* SDMMC operations FPs are the element of structure*/
+static const struct sdhc_cdns_ops *cdns_sdmmc_ops;
+
+struct sdhc_cdns_config {
+	DEVICE_MMIO_NAMED_ROM(reg_base);
+	DEVICE_MMIO_NAMED_ROM(combo_phy);
+	/* Clock rate for host */
+	uint32_t clk_rate;
+	/* power delay prop for host */
+	uint32_t power_delay_ms;
+	/* run time device structure */
+	const struct device *cdns_clk_dev;
+	/* type to identify a clock controller sub-system */
+	clock_control_subsys_t clkid;
+	/* Reset controller device configuration. */
+	const struct reset_dt_spec reset_sdmmc;
+	const struct reset_dt_spec reset_sdmmcocp;
+	const struct reset_dt_spec reset_softphy;
+};
+
+struct sdhc_cdns_data {
+	DEVICE_MMIO_NAMED_RAM(reg_base);
+	DEVICE_MMIO_NAMED_RAM(combo_phy);
+	/* Host controller parameters */
+	struct sdhc_cdns_params params;
+	/* sdmmc device informartaion for host */
+	struct sdmmc_device_info info;
+	/* Input/Output configuration */
+	struct sdhc_io host_io;
+};
+
+static int sdhc_cdns_request(const struct device *dev,
+	struct sdhc_command *cmd,
+	struct sdhc_data *data)
+{
+	int ret = 0;
+	struct sdmmc_cmd cdns_sdmmc_cmd;
+
+	if (cmd == NULL) {
+		LOG_ERR("Wrong CMD parameter");
+		return -EINVAL;
+	}
+
+	/* Initialization of command structure */
+	cdns_sdmmc_cmd.cmd_idx =  cmd->opcode;
+	cdns_sdmmc_cmd.cmd_arg = cmd->arg;
+	cdns_sdmmc_cmd.resp_type = (cmd->response_type & SDHC_NATIVE_RESPONSE_MASK);
+
+	/* Sending command as per the data or non data */
+	if (data) {
+		ret = cdns_sdmmc_ops->prepare(data->block_addr, (uintptr_t)data->data,
+				data);
+		if (ret != 0) {
+			LOG_ERR("DMA Prepare failed");
+			return -EINVAL;
+		}
+	}
+
+	ret = cdns_sdmmc_ops->send_cmd(&cdns_sdmmc_cmd, data);
```

Comment:
Right now, the Cadence IP6061 driver bare metal functions are abstracted in sdhc_cdns_ll.c files and as of now all these function pointers are pointing to single function since only this IP supported. In future, if any new Cadence IP from the same family is introduced then we will make use of these function pointers to point different sdhc_cdns_ll.c file and by keeping sdhc_cdns.c wrapper file same across both the drivers.

Reply:
Understood. If this abstraction layer will make porting future IP to Zephyr easier than I am ok with it. Just to clarify- is the sdhc_cdns_ll.c a wrapper that exists for other IP not presently in Zephyr? Or is this wrapper being designed specifically for Zephyr support?

Reply:
This is designed specially for zephyr support.

Reply:
@roymurlidhar do you expect there to be multiple cadence SDHCs of different IP types/versions on the same platform? because that would be the main justification for function pointers instead of just compiling different definitions for these functions

Reply:
also, if this is all zephyr specific/native anyways, what benefit do you get from having another abstraction layer, when you can write just a different zephyr driver with one layer for the different IP

Reply:
@danieldegrasse @decsny 

> also, if this is all zephyr specific/native anyways, what benefit do you get from having another abstraction layer, when you can write just a different zephyr driver with one layer for the different IP
> 
> a lot of the functions you're implementing in the 2nd layer are 1 to 1 with the sdhc api that already exists and you are implementing, for example your reset function implementation of the zephyr api is just calling your 2nd layer reset api and doing nothing else, what is the purpose of this, this is why there is already a first layer api for this, because different devices have different implementations for it.

Hello @decsny,

Agree with your comments. However, our main objective of splitting this driver into 2 layers is:
1st layer of the driver will handle all the Zephyr device driver framework related code
2nd layer will have the bare metal code (Independent of Zephyr or any other OS) for all the functionalities of the IP.

The motive behind this abstraction is: to provide Zephyr driver and as well as bare metal code also to our (Intel Corporation) customers.

Reply:
> Ok. For my own understanding- is this abstraction layer going to enable reuse in the future? IE would a future driver for different IP possibly reuse functions from the sdhc_cdns_ll.c file? I'm trying to understand what the benefit/reasoning is for introducing this additional layer of abstraction

Hello @danieldegrasse,

I quoted our main objective of this abstraction layer to @decsny reply.
Yes, if the changes are minimal, we can re-use the same abstraction layer for different Cadence IP from the same family.

Thank you.

Reply:
@gdengi thank you for the clarification. Provided that this layer can be reused within Zephyr, I am ok with the approach.

Reply:
Thank you @danieldegrasse for your valuable feedbacks and review comments.

Reply:
@decsny @danieldegrasse Is there any feedback or comments. If not can we approve this PR.